prime-router/docs/proposals/0013-batch-retry.md

```diff
@@ -0,0 +1,66 @@
+on the queue to be handled immediately by the batchFunction. This is a minimal-change approach that leaves the
+Task table management as it is but severely reduces our performance bottlenecks and race conditions. Once done, the
+single run of the batch function for that receiver will pull either all outstanding records (or up to <limit> if there)
+is one for this receiver and batch them.
+
+### Nifty Warehouse Metaphor
+Right now we have a front desk person who is getting paper reports - people are dropping by and sticking them in the
+inbox at any time. The front desk person is copying incoming reports as needed, adding a few stamps, and for each 
+report is calling the mail room to say 'hey, got a report for Receiver X, come get it ready to ship at 
+noon' and then sticking it in the internal folder for that receiver. At noon the mail room sends one person per call 
+they got to get files ready to send to the receiver. Each of these people takes the time to get up from their desk and 
+walk over - the first person to get there for each receiver grabs them all out of the folder, puts them in an
+envelope, and sticks it in the outbox for the receiver. The rest of them show up, grumble that there is nothing to do
+and go back to their desk.
+
+In the proposed changes the front desk person still gets the reports and puts them in a folder, but instead of calling
+the mail room each time they get a report there is a front desk assistant that keeps track of schedules and receivers.
+At noon, they call over to the mail room to say 'got some reports for Receiver X, please send someone over'. One person
+walks over, packages up the reports, and gets to them to appropriate outbox. If for some reason they are unable to 
+do that (tripped on their shoelaces or something) the assistant can call back to the mail room to ask them to send
+someone else.
+
+### Dev Notes
+To implement these changes, we would need to
+* determine minimal batching granularity
+* create a cron job that runs at that granularity
+* either cache or dynamically build a directory of which receivers get batches at which time
+* when the cron job triggers, add a 'batch' queue message with the name of the receiver
+* rename 'ReceiverEvent' to 'BatchEvent' for clarity (in the codebase - Event.kt)
+* look into the 'at' parameter in batchFunction to determine if it is still needed
+* we would need to look into a setting change for receiver to indicate 'sendIfEmptyBatch' and generate empty batch
+files if set to true
+* Since the batchFunction locks the task rows that it is processing, it doesn't matter if runs for the same receiver
+overlap (this is not ideal, but still a far cry better than what we have today)
+
+### Benefits Gained / Problems Addressed
+- Scalability - one queue message per receiver per batch bucket
+- Recoverabilty/Retry - if a batch fails, the queue message can be re-added to the queue and re-processed
```

Comment:
In this case both will happen. An initial retry for some very transient error, and the next time the cron triggers a more robust recovery.

Reply:
The 'past-due' style combines retry and batch into the same mechanism, instead of having to code a figure-out-the-batch-time, then separately code a retry mechanism.

However, the 'past due' style does not handle the sendIfEmptyBatch feature as elegantly.

Reply:
yup we're saying the same thing. The queue message is just the trigger for the batch function, the batch function will always get all past due messages.

Reply:
Just to be clear, in order to get past-due messages, it uses the next_action_at field in TASK?

(Oh, and to answer an earlier question: for sanity's sake, yes let's assume it's never the case that a report should be held for future batching)